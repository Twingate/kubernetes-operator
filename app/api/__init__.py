# ruff: noqa: F401
from .client import TwingateAPIClient
