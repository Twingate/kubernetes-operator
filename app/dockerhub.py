from collections.abc import Iterator

import requests
from semantic_version import NpmSpec, Version

_DOCKER_HUB_API_BASE_URL = "https://hub.docker.com/v2"
_DOCKER_HUB_TAGS_API_URL = (
    f"{_DOCKER_HUB_API_BASE_URL}/repositories/twingate/kubernetes-operator/tags"
)


def get_all_operator_tags() -> Iterator[str]:
    url = f"{_DOCKER_HUB_TAGS_API_URL}?page_size=100"
    response = requests.get(url, timeout=3)
    data = response.json()
    for result in data["results"]:
        yield result["name"]


def get_all_operator_semver_tags(
    *, allow_prerelease: bool = False
) -> Iterator[Version]:
    for tag in get_all_operator_tags():
        try:
            v = Version(tag)
            if v.prerelease and not allow_prerelease:
                continue

            yield v
        except ValueError:
            continue


def get_all_operator_semver_tags_by_specifier(
    specifier: str, *, allow_prerelease: bool = False
) -> Iterator[Version]:
    specifier = NpmSpec(specifier)
    return specifier.filter(
        get_all_operator_semver_tags(allow_prerelease=allow_prerelease)
    )


def get_latest(specifier: str, *, allow_prerelease: bool = False) -> Version | None:
    matched_versions = list(
        get_all_operator_semver_tags_by_specifier(
            specifier, allow_prerelease=allow_prerelease
        )
    )
    return max(matched_versions)
