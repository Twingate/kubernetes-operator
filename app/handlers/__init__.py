# ruff: noqa: F403,F403
from .handlers_resource import *
from .handlers_resource_access import *
