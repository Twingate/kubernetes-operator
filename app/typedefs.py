import logging
from typing import TYPE_CHECKING, Any, Union

if TYPE_CHECKING:
    LoggerAdapter = logging.LoggerAdapter[Any]
else:
    LoggerAdapter = logging.LoggerAdapter

Logger = Union[logging.Logger, LoggerAdapter]
