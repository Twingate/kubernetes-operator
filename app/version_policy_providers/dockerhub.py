from collections.abc import Iterator

import requests

from app.version_policy_providers.base import VersionPolicyProvider


class DockerhubVersionPolicyProvider(VersionPolicyProvider):
    _DOCKER_HUB_API_BASE_URL = "https://hub.docker.com/v2"

    def __init__(self, _repository: str | None = None):
        self.repository = "twingate/connector"
        self.tags_api_url = (
            f"{self._DOCKER_HUB_API_BASE_URL}/repositories/{self.repository}/tags"
        )

    def __call(self, page: int = 0):
        url = f"{self.tags_api_url}?page_size=100&page={page}"
        response = requests.get(url, timeout=6)
        return response.json()

    def __get_all_api_pages(self):
        page = 0
        while True:
            data = self.__call(page)
            yield data
            if not data["next"]:
                break
            page += 1

    def get_all_tags(self) -> Iterator[str]:
        for data in self.__get_all_api_pages():
            for result in data["results"]:
                yield result["name"]
