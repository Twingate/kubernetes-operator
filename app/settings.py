import binascii
import logging
import os
import tomllib
from base64 import b64decode
from typing import Annotated, ClassVar

from pydantic.functional_validators import AfterValidator
from pydantic_core._pydantic_core import ValidationError
from pydantic_settings import BaseSettings, SettingsConfigDict


def validate_graphql_global_id(value: str) -> str:
    value_decoded: str = ""
    try:
        value_bytes: bytes = value.encode("ascii")
        value_decoded = b64decode(value_bytes).decode("ascii")
    except (binascii.Error, UnicodeDecodeError, UnicodeEncodeError):
        pass

    if ":" not in value_decoded:
        raise ValueError("Invalid global id")

    return value


GlobalID = Annotated[str, AfterValidator(validate_graphql_global_id)]


class TwingateOperatorSettings(BaseSettings):
    model_config = SettingsConfigDict(env_prefix="TWINGATE_")
    NULL_RN_ID: ClassVar[str] = "ZmFrZTp2YWx1ZQo="  # temp value ("fake:value" encoded)

    api_key: str
    network: str
    remote_network_id: GlobalID = NULL_RN_ID
    remote_network_name: str | None = None
    host: str = "twingate.com"

    @property
    def full_url(self) -> str:
        return f"https://{self.network}.{self.host}"

    def __init__(self, *args, **kwargs):
        from app.api import TwingateAPIClient

        super().__init__(*args, **kwargs)

        if self.remote_network_name:
            client = TwingateAPIClient(self)
            rn = client.get_remote_network_by_name(self.remote_network_name)
            if not rn:
                raise ValueError(f"Remote network {self.remote_network_name} not found")
            self.remote_network_id = rn.id

        if self.remote_network_id == self.NULL_RN_ID:
            raise ValidationError("Remote network id is required")


__settings: TwingateOperatorSettings | None = None
__version: str | None = None


def get_settings() -> TwingateOperatorSettings:  # pragma: no cover
    global __settings
    if not __settings:
        __settings = TwingateOperatorSettings()
    return __settings


def get_version() -> str:  # pragma: no cover
    # ruff: noqa: E722
    global __version
    if not __version:
        try:
            filename = os.path.join(os.path.dirname(__file__), "..", "pyproject.toml")
            with open(filename, "rb") as f:
                data = tomllib.load(f)
                __version = (
                    data.get("tool", {}).get("poetry", {}).get("version", "0.0.0")
                )
        except:
            logging.exception("Failed to load version from pyproject.toml")
            __version = "0.0.0"

    return __version or "0.0.0"
