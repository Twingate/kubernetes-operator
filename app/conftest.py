from unittest.mock import patch

import pytest


@pytest.fixture(scope="module", autouse=True)
def _mock_settings():
    from app.settings import TwingateOperatorSettings

    settings = TwingateOperatorSettings(
        api_key="apikey",
        network="network",
        remote_network_id="UmVtb3RlTmV0d29yazoxMjMK",
    )

    with patch("app.crds.get_settings", return_value=settings):
        yield
