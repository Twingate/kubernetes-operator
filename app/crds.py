import logging
from collections.abc import MutableMapping
from datetime import datetime
from enum import Enum
from typing import Any

import kubernetes.client
import pendulum
from croniter import croniter
from pydantic import (
    BaseModel,
    ConfigDict,
    Field,
    ValidationInfo,
    field_validator,
    model_validator,
)
from pydantic.alias_generators import to_camel
from semantic_version import NpmSpec

from app.settings import get_settings
from app.version_policy_providers import get_provider

K8sObject = MutableMapping[Any, Any]
OptionalK8sObject = K8sObject | None


class K8sMetadata(BaseModel):
    model_config = ConfigDict(
        frozen=True, populate_by_name=True, extra="allow", alias_generator=to_camel
    )

    uid: str
    name: str
    namespace: str

    @property
    def owner_reference_object(self) -> dict:
        return {
            "apiVersion": "twingate.com/v1",
            "kind": "TwingateResource",
            "name": self.name,
            "uid": self.uid,
        }


class BaseK8sModel(BaseModel):
    model_config = ConfigDict(
        frozen=True, populate_by_name=True, extra="allow", alias_generator=to_camel
    )

    api_version: str
    kind: str
    metadata: K8sMetadata
    status: dict[str, Any] | None = None


# region TwingateResourceCRD


class ProtocolPolicy(str, Enum):
    ALLOW_ALL = "ALLOW_ALL"
    RESTRICTED = "RESTRICTED"


class ProtocolRange(BaseModel):
    model_config = ConfigDict(
        frozen=True, populate_by_name=True, alias_generator=to_camel
    )

    start: int = Field(ge=1, le=65535)
    end: int = Field(ge=1, le=65535)

    @model_validator(mode="after")
    def check_ports(self):
        if self.start > self.end:
            raise ValueError("Start port value must be less or equal to end port value")

        return self


class ResourceProtocol(BaseModel):
    model_config = ConfigDict(
        frozen=True, populate_by_name=True, alias_generator=to_camel
    )

    policy: ProtocolPolicy = ProtocolPolicy.ALLOW_ALL
    ports: list[ProtocolRange] = Field(default_factory=list)

    @model_validator(mode="after")
    def check_policy_ports(self):
        if self.policy == ProtocolPolicy.ALLOW_ALL and self.ports:
            raise ValueError("ports can't be set if policy is ALLOW_ALL")

        return self


class ResourceProtocols(BaseModel):
    model_config = ConfigDict(
        frozen=True, populate_by_name=True, alias_generator=to_camel
    )

    allow_icmp: bool | None = True
    tcp: ResourceProtocol = Field(default_factory=ResourceProtocol)
    udp: ResourceProtocol = Field(default_factory=ResourceProtocol)


class ResourceSpec(BaseModel):
    model_config = ConfigDict(
        frozen=True, populate_by_name=True, alias_generator=to_camel
    )

    id: str | None = None
    name: str
    address: str
    alias: str | None = None
    remote_network_id: str = Field(
        default_factory=lambda: get_settings().remote_network_id
    )
    security_policy_id: str | None = None
    is_visible: bool = True
    is_browser_shortcut_enabled: bool = True
    protocols: ResourceProtocols = Field(default_factory=ResourceProtocols)

    def __is_wildcard(self):
        return "*" in self.address or "?" in self.address

    @model_validator(mode="after")
    def check_address_if_is_browser_shortcut_enabled(self):
        if self.is_browser_shortcut_enabled and self.__is_wildcard():
            raise ValueError(
                "isBrowserShortcutEnabled can't be True for wildcard addresses"
            )

        return self


class TwingateResourceCRD(BaseK8sModel):
    model_config = ConfigDict(frozen=True, populate_by_name=True, extra="allow")

    spec: ResourceSpec


# endregion

# region TwingateResourceAccessCRD


class PrincipalTypeEnum(str, Enum):
    Group = "group"
    ServiceAccount = "serviceAccount"


class _ResourceRef(BaseModel):
    name: str
    namespace: str = Field(default="default")


class _PrincipalExternalRef(BaseModel):
    model_config = ConfigDict(populate_by_name=True, alias_generator=to_camel)

    type: PrincipalTypeEnum
    name: str


class ResourceAccessSpec(BaseModel):
    model_config = ConfigDict(populate_by_name=True, alias_generator=to_camel)

    resource_ref: _ResourceRef
    principal_id: str | None = None
    principal_external_ref: _PrincipalExternalRef | None = None
    security_policy_id: str | None = None

    @model_validator(mode="after")
    def validate_princiapl_id_or_principal_external_ref(self):
        if self.principal_id or self.principal_external_ref:
            return self

        raise ValueError("Missing principal_id or principal_external_ref")

    @property
    def resource_ref_fullname(self) -> str:
        return f"{self.resource_ref.namespace}/{self.resource_ref.name}"

    def get_resource_ref_object(self) -> OptionalK8sObject:
        try:
            kapi = kubernetes.client.CustomObjectsApi()
            response = kapi.get_namespaced_custom_object(
                "twingate.com",
                "v1beta",
                self.resource_ref.namespace,
                "twingateresources",
                self.resource_ref.name,
            )
            logging.info("ResourceAccessCRD.get_resource_ref_object got: %s", response)
            return response
        except kubernetes.client.exceptions.ApiException as api_ex:
            if api_ex.status == 404:
                logging.warning(
                    "ResourceAccessCRD.get_resource_ref_object: resource not found."
                )
            else:
                logging.exception("ResourceAccessCRD.get_resource_ref_object failed")

            return None

    def get_resource(self) -> TwingateResourceCRD | None:
        resource_ref_object = self.get_resource_ref_object()
        if not resource_ref_object:
            return None

        resource = ResourceSpec(**resource_ref_object.pop("spec"))
        metadata = K8sMetadata(**resource_ref_object.pop("metadata"))
        return TwingateResourceCRD(
            metadata=metadata, spec=resource, **resource_ref_object
        )


class TwingateResourceAccessCRD(BaseK8sModel):
    model_config = ConfigDict(frozen=True, populate_by_name=True, extra="allow")

    spec: ResourceAccessSpec


# endregion

# region TwingateConnector


class ConnectorImagePolicyProvidersEnum(str, Enum):
    dockerhub = "dockerhub"
    google = "google"


class ConnectorImagePolicy(BaseModel):
    model_config = ConfigDict(
        frozen=True, populate_by_name=True, alias_generator=to_camel, extra="allow"
    )

    provider: ConnectorImagePolicyProvidersEnum = (
        ConnectorImagePolicyProvidersEnum.dockerhub
    )
    repository: str = "twingate/connector"
    schedule: str = "0 0 * * *"
    version: str = "^1.0.0"
    allow_prerelease: bool = False

    @field_validator("version")
    @classmethod
    def check_valid_version_specifier(cls, v: str, info: ValidationInfo) -> str:
        try:
            NpmSpec(v)
        except ValueError as vex:
            raise ValueError("Invalid version specifier") from vex
        return v

    @field_validator("schedule")
    @classmethod
    def check_valid_crontab(cls, v: str, info: ValidationInfo) -> str:
        if v:
            try:
                croniter(v)
            except ValueError as vex:
                raise ValueError("Invalid schedule value") from vex
        return v

    def get_next_date_iso8601(self) -> str:
        next_date = croniter(self.schedule, pendulum.now("UTC")).get_next(datetime)
        return pendulum.instance(next_date).to_iso8601_string()

    def get_image(self) -> str:
        provider = get_provider(self.provider.value, repository=self.repository)
        if tag := provider.get_latest(
            self.version, allow_prerelease=self.allow_prerelease
        ):
            return f"{self.repository}:{tag}"

        raise ValueError(
            f"Could not find valid tag for '{self.version}' at '{self.repository}'"
        )


class ConnectorImage(BaseModel):
    model_config = ConfigDict(
        frozen=True, populate_by_name=True, alias_generator=to_camel, extra="allow"
    )

    repository: str = "twingate/connector"
    tag: str = "1"

    def __str__(self):
        return f"{self.repository}:{self.tag}"


class ConnectorSpec(BaseModel):
    model_config = ConfigDict(
        frozen=True, populate_by_name=True, alias_generator=to_camel, extra="allow"
    )

    id: str | None = None
    name: str | None = None
    log_level: int = 3
    has_status_notifications_enabled: bool = True
    image: ConnectorImage | None = None
    image_policy: ConnectorImagePolicy | None = None
    container_extra: dict[str, Any] = {}
    pod_extra: dict[str, Any] = {}
    pod_annotations: dict[str, Any] = {}

    remote_network_id: str = Field(
        default_factory=lambda: get_settings().remote_network_id
    )

    @model_validator(mode="after")
    def validate_image_or_image_policy(self):
        if self.image or self.image_policy:
            return self

        # Default to having `image`
        return self.model_copy(update=dict(image=ConnectorImage()))

    def get_image(self) -> str:
        if image := self.image:
            return str(image)

        if image_policy := self.image_policy:
            return image_policy.get_image()

        # Impossible to get here because of our model validator
        raise ValueError("Invalid ConnectorSpec")  # pragma: no cover


class TwingateConnectorCRD(BaseK8sModel):
    model_config = ConfigDict(frozen=True, populate_by_name=True, extra="allow")

    spec: ConnectorSpec


# endregion
