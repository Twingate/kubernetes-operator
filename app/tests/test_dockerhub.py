import os
import pathlib
from unittest.mock import patch

import pytest
from semantic_version import Version

from app.dockerhub import (
    _DOCKER_HUB_TAGS_API_URL,
    get_all_operator_semver_tags,
    get_all_operator_semver_tags_by_specifier,
    get_all_operator_tags,
    get_latest,
)

VERSIONS_LIST = [
    "latest",
    "dev",
    "1.1.2-dev.6722290577",
    "1.1.2",
    "1.1",
    "0.1.2-dev.6722290577",
    "0.1.2",
    "0.1",
    "0",
    "0.1.2-dev.6716132391",
    "0.1.2-dev.6716010709",
    "0.1.1-dev.6702190178",
    "0.1.1",
    "0.1.1-dev.6702103173",
    "0.1.0-dev.6698032918",
    "0.1.0",
    "0.1.0-dev.6697571821",
    "0.1.0-dev.6697098399",
    "0.1.0-dev.6697032623",
    "0.1.0-dev.6686861896",
    "0.1.0-dev.6683812832",
    "0.1.0-dev.6672988848",
]


@pytest.fixture(scope="session")
def dockerhub_image_tags_sample_response():
    dir_path = os.path.dirname(os.path.realpath(__file__))
    file = pathlib.Path(dir_path, "./test_data/dockerhub_tags_response.json")
    with open(file) as f:
        data = f.read()
    return data


def test_get_all_operator_tags(
    mocked_responses, snapshot, dockerhub_image_tags_sample_response
):
    mocked_responses.get(
        _DOCKER_HUB_TAGS_API_URL, status=200, body=dockerhub_image_tags_sample_response
    )
    actual = list(get_all_operator_tags())
    assert actual == snapshot


def test_get_operator_semver_tags(snapshot):
    with patch("app.dockerhub.get_all_operator_tags", return_value=VERSIONS_LIST):
        actual = list(get_all_operator_semver_tags())
        assert actual == snapshot


def test_get_operator_semver_tags_prerelease(snapshot):
    with patch("app.dockerhub.get_all_operator_tags", return_value=VERSIONS_LIST):
        actual = list(get_all_operator_semver_tags(allow_prerelease=True))
        assert actual == snapshot


def test_get_all_operator_semver_tags_by_specifier(snapshot):
    with patch("app.dockerhub.get_all_operator_tags", return_value=VERSIONS_LIST):
        actual = list(get_all_operator_semver_tags_by_specifier("0.1.x"))
        assert actual == snapshot


def test_get_latest():
    with patch("app.dockerhub.get_all_operator_tags", return_value=VERSIONS_LIST):
        actual = get_latest("0.1.x")
        assert actual == Version("0.1.2")

        actual = get_latest("^1.0.0")
        assert actual == Version("1.1.2")
