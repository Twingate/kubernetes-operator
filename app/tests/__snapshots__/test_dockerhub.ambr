# serializer version: 1
# name: test_get_all_operator_semver_tags_by_specifier
  list([
    Version('0.1.2'),
    Version('0.1.1'),
    Version('0.1.0'),
  ])
# ---
# name: test_get_all_operator_tags
  list([
    'latest',
    'dev',
    '0.1.2-dev.6722290577',
    '0.1.2',
    '0.1',
    '0',
    '0.1.2-dev.6716132391',
    '0.1.2-dev.6716010709',
    '0.1.1-dev.6702190178',
    '0.1.1',
    '0.1.1-dev.6702103173',
    '0.1.0-dev.6698032918',
    '0.1.0',
    '0.1.0-dev.6697571821',
    '0.1.0-dev.6697098399',
    '0.1.0-dev.6697032623',
    '0.1.0-dev.6686861896',
    '0.1.0-dev.6683812832',
    '0.1.0-dev.6672988848',
  ])
# ---
# name: test_get_operator_semver_tags
  list([
    Version('1.1.2'),
    Version('0.1.2'),
    Version('0.1.1'),
    Version('0.1.0'),
  ])
# ---
# name: test_get_operator_semver_tags_prerelease
  list([
    Version('1.1.2-dev.6722290577'),
    Version('1.1.2'),
    Version('0.1.2-dev.6722290577'),
    Version('0.1.2'),
    Version('0.1.2-dev.6716132391'),
    Version('0.1.2-dev.6716010709'),
    Version('0.1.1-dev.6702190178'),
    Version('0.1.1'),
    Version('0.1.1-dev.6702103173'),
    Version('0.1.0-dev.6698032918'),
    Version('0.1.0'),
    Version('0.1.0-dev.6697571821'),
    Version('0.1.0-dev.6697098399'),
    Version('0.1.0-dev.6697032623'),
    Version('0.1.0-dev.6686861896'),
    Version('0.1.0-dev.6683812832'),
    Version('0.1.0-dev.6672988848'),
  ])
# ---
